python/paddle/fluid/contrib/mixed_precision/fp16_utils.py

```diff
@@ -0,0 +1,99 @@
+#   Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+from ... import core
+from ... import layers
+from ... import framework
+
+
+def cast_fp16_to_fp32(i, o, prog):
+    prog.global_block().append_op(
+        type="cast",
+        inputs={"X": i},
+        outputs={"Out": o},
+        attrs={
+            "in_dtype": core.VarDesc.VarType.FP16,
+            "out_dtype": core.VarDesc.VarType.FP32
+        })
+
+
+def cast_fp32_to_fp16(i, o, prog):
+    prog.global_block().append_op(
+        type="cast",
+        inputs={"X": i},
+        outputs={"Out": o},
+        attrs={
+            "in_dtype": core.VarDesc.VarType.FP32,
+            "out_dtype": core.VarDesc.VarType.FP16
+        })
+
+
+def copy_to_master_param(p, block):
+    v = block.vars.get(p.name, None)
+    if v is None:
+        raise ValueError("no param name %s found!" % p.name)
+    new_p = framework.Parameter(
+        block=block,
+        shape=v.shape,
+        dtype=core.VarDesc.VarType.FP32,
```

Comment:
Why the dtype must be `FP32` here? Is it possible that the dtype is `fp64`?

Reply:
In principle, the master parameter can be `float64`. But there are some [hard-coded implementations](https://github.com/PaddlePaddle/Paddle/blob/1202d3fc740b065590f93a6029ab2126ca9dafff/python/paddle/fluid/layers/nn.py#L3038), and the `fp64` support seems not to be that straightforward. So we are going to only support `float32` temporarily because it is more common used. Maybe we can go back to consider `fp64` some day in the future.